docs/enhancements/scalable-syncset-processing.md

```diff
@@ -0,0 +1,91 @@
+# Scalable SyncSet Processing
+
+## Summary
+
+In large Hive deployments (current recommendation is to stay below 1000 ClusterDeployments), with fairly heavy SyncSet usage (approximately 60 SyncSets with about 10 resources each, a total of 600 resources to apply), the SyncSet processing has proven to be our only relevant performance bottleneck. This proposal is to implement a system by which we can horizontally scale a set of pods for SyncSet processing to eliminate the bottleneck instead of continuing to fight with it.
+
+## Motivation
+
+Even after implementing dramatic changes to improve SyncSet performance, we are struggling to get past a max of about 20 applies/s. Full clustersync controller reconcile time can range from 2 to 5 minutes to process everything for one cluster under load. At 20 goroutines on an AWS c5.4xlarge, the workqueue is struggling to process.
+
+Individual apply time appears to be approximately 0.2s - 1.6s, although 99% of all applys complete in under 0.5s. 0.5s * 600 = 300 seconds to process an entire cluster.
+
+Attempts to increase the goroutines to 100 lead to pod instability, at present problems with allocating OS threads.
+
+Moving to a design where we can horizontally scale the number of pods processing syncsets should let us scale out to meet the requirements of heavier SyncSet usage. In doing so we may be able to push far beyond the 1k ClusterDeployment limit and possibly save substantial cost for extremely large Hive deployments now requiring fewer shards.
```

Comment:
My recollection is that the 1k limit was based on the size of the ClusterSync resources. If that is correct, then horizontally scaling will still not allow Hive to move beyond the 1k limit.